procrastinate/sql/migrations/02.00.03_01_add_job_priority.sql

```diff
@@ -1,5 +1,7 @@
 ALTER TABLE procrastinate_jobs ADD COLUMN priority integer DEFAULT 0 NOT NULL;
 
+CREATE UNIQUE INDEX idx_procrastinate_jobs_priority_id ON procrastinate_jobs (priority, id) WHERE status = 'todo';
```

Comment:
Excellent idea! Did you benchmark in some way?

Reply:
Nope, no benchmarking ... just divine inspiration 😂. If there are many jobs, many of them have to be sorted by priority. So, an index will help. I guess, in many cases, it won't make any difference; on the other hand, it's only a small index.

Reply:
Ah, then can I ask you to `EXPLAIN ANALYZE SELECT procrastinate_defer_job()` and check that the index is actually used?

There are already other indexes, maybe what we need is to modify an existing index rather than create new ones ?

Reply:
Sure, makes sense!👍

Reply:
Ok, this is what I get:

```
procrastinate=# EXPLAIN ANALYZE SELECT id, status, task_name, priority, lock, queueing_lock, args, scheduled_at, queue_name, attempts FROM procrastinate_fetch_job(ARRAY['index']);
                                                       QUERY PLAN                                                       
------------------------------------------------------------------------------------------------------------------------
 Function Scan on procrastinate_fetch_job  (cost=0.25..0.26 rows=1 width=672) (actual time=0.395..0.395 rows=1 loops=1)
 Planning Time: 0.037 ms
 Execution Time: 0.408 ms
(3 rows)
```

It seems that this index isn't used at all 😳. Does this mean it is simply not necessary? Or is it wrongly defined? I think I need your help here 🙏.

Reply:
I did some further experiments. Here are some results with about 150,000 jobs waiting to be processed and a random priority for each job.

```
procrastinate=# EXPLAIN ANALYZE SELECT id, status, task_name, priority, lock, queueing_lock, args, scheduled_at, queue_name, attempts FROM procrastinate_fetch_job(ARRAY['index']);
                                                         QUERY PLAN                                                         
----------------------------------------------------------------------------------------------------------------------------
 Function Scan on procrastinate_fetch_job  (cost=0.25..0.26 rows=1 width=672) (actual time=128.042..128.043 rows=1 loops=1)
 Planning Time: 0.042 ms
 Execution Time: 128.057 ms
(3 rows)
```

It is still quite fast, even without an index that is being used. So should we just go without any additional index?

Reply:
Seeing your results, I'd say you'd have more details if you ran the explain on the query that is ran by the procedure rather than running the procedure itself so:

```
EXPLAIN ANALYZE SELECT jobs.*
FROM procrastinate_jobs AS jobs
WHERE
    -- reject the job if its lock has earlier jobs
    NOT EXISTS (
        SELECT 1
            FROM procrastinate_jobs AS earlier_jobs
            WHERE
                jobs.lock IS NOT NULL
                AND earlier_jobs.lock = jobs.lock
                AND earlier_jobs.status IN ('todo', 'doing')
                AND earlier_jobs.id < jobs.id)
    AND jobs.status = 'todo'
    AND (target_queue_names IS NULL OR jobs.queue_name = ANY( target_queue_names ))
    AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())
ORDER BY jobs.priority DESC, jobs.id ASC LIMIT 1
```

Also, feel free to try with 10x more jobs (or more) if it helps make the effect of the index visible

Reply:
Ok, here are some more results:

For 1.000.000 jobs
```
procrastinate=# EXPLAIN ANALYZE SELECT id, status, task_name, priority, lock, queueing_lock, args, scheduled_at, queue_name, attempts FROM procrastinate_fetch_job(ARRAY['index']);
                                                        QUERY PLAN                                                        
--------------------------------------------------------------------------------------------------------------------------
 Function Scan on procrastinate_fetch_job  (cost=0.25..0.26 rows=1 width=672) (actual time=95.765..95.766 rows=1 loops=1)
 Planning Time: 0.040 ms
 Execution Time: 95.780 ms
(3 rows)
```

```
procrastinate=# EXPLAIN ANALYZE SELECT jobs.*                                                                                                                                      
            FROM procrastinate_jobs AS jobs
            WHERE
                -- reject the job if its lock has earlier jobs
                NOT EXISTS (
                    SELECT 1
                        FROM procrastinate_jobs AS earlier_jobs
                        WHERE
                            jobs.lock IS NOT NULL
                            AND earlier_jobs.lock = jobs.lock
                            AND earlier_jobs.status IN ('todo', 'doing')
                            AND earlier_jobs.id < jobs.id)
                AND jobs.status = 'todo'
                AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())
            ORDER BY jobs.priority DESC, jobs.id ASC LIMIT 1;
			
QUERY PLAN                                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=11885.73..11885.85 rows=1 width=191) (actual time=54.537..59.374 rows=1 loops=1)
   ->  Gather Merge  (cost=11885.73..22189.50 rows=89598 width=191) (actual time=54.536..59.371 rows=1 loops=1)
         Workers Planned: 1
         Workers Launched: 1
         ->  Sort  (cost=10885.72..11109.72 rows=89598 width=191) (actual time=52.067..52.070 rows=1 loops=2)
               Sort Key: jobs.priority DESC, jobs.id
               Sort Method: top-N heapsort  Memory: 25kB
               Worker 0:  Sort Method: top-N heapsort  Memory: 25kB
               ->  Parallel Hash Anti Join  (cost=4548.28..10437.73 rows=89598 width=191) (actual time=11.318..34.528 rows=76164 loops=2)
                     Hash Cond: (jobs.lock = earlier_jobs.lock)
                     Join Filter: ((jobs.lock IS NOT NULL) AND (earlier_jobs.id < jobs.id))
                     ->  Parallel Seq Scan on procrastinate_jobs jobs  (cost=0.00..4534.28 rows=89598 width=191) (actual time=0.013..14.251 rows=76164 loops=2)
                           Filter: ((status = 'todo'::procrastinate_job_status) AND ((scheduled_at IS NULL) OR (scheduled_at <= now())))
                           Rows Removed by Filter: 10
                     ->  Parallel Hash  (cost=3754.97..3754.97 rows=63465 width=40) (actual time=10.003..10.004 rows=76174 loops=2)
                           Buckets: 262144  Batches: 1  Memory Usage: 2048kB
                           ->  Parallel Index Only Scan using procrastinate_jobs_id_lock_idx on procrastinate_jobs earlier_jobs  (cost=0.42..3754.97 rows=63465 width=40) (actual time=0.027..7.096 rows=76174 loops=2)
                                 Heap Fetches: 53
 Planning Time: 0.444 ms
 Execution Time: 59.413 ms
(20 rows)
```

And also 2.000.000 jobs
```
procrastinate=# EXPLAIN ANALYZE SELECT id, status, task_name, priority, lock, queueing_lock, args, scheduled_at, queue_name, attempts FROM procrastinate_fetch_job(ARRAY['index']);
                                                        QUERY PLAN                                                        
--------------------------------------------------------------------------------------------------------------------------
 Function Scan on procrastinate_fetch_job  (cost=0.25..0.26 rows=1 width=672) (actual time=97.055..97.055 rows=1 loops=1)
 Planning Time: 0.050 ms
 Execution Time: 97.070 ms
(3 rows)
```

```
procrastinate=# EXPLAIN ANALYZE SELECT jobs.*                                                                                                                                      
            FROM procrastinate_jobs AS jobs
            WHERE
                -- reject the job if its lock has earlier jobs
                NOT EXISTS (
                    SELECT 1
                        FROM procrastinate_jobs AS earlier_jobs
                        WHERE
                            jobs.lock IS NOT NULL
                            AND earlier_jobs.lock = jobs.lock
                            AND earlier_jobs.status IN ('todo', 'doing')
                            AND earlier_jobs.id < jobs.id)
                AND jobs.status = 'todo'
                AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())
            ORDER BY jobs.priority DESC, jobs.id ASC LIMIT 1;
			

QUERY PLAN                                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=11885.73..11885.85 rows=1 width=191) (actual time=67.032..71.506 rows=1 loops=1)
   ->  Gather Merge  (cost=11885.73..22189.50 rows=89598 width=191) (actual time=67.031..71.504 rows=1 loops=1)
         Workers Planned: 1
         Workers Launched: 1
         ->  Sort  (cost=10885.72..11109.72 rows=89598 width=191) (actual time=64.441..64.444 rows=1 loops=2)
               Sort Key: jobs.priority DESC, jobs.id
               Sort Method: top-N heapsort  Memory: 25kB
               Worker 0:  Sort Method: top-N heapsort  Memory: 25kB
               ->  Parallel Hash Anti Join  (cost=4548.28..10437.73 rows=89598 width=191) (actual time=14.048..43.027 rows=76164 loops=2)
                     Hash Cond: (jobs.lock = earlier_jobs.lock)
                     Join Filter: ((jobs.lock IS NOT NULL) AND (earlier_jobs.id < jobs.id))
                     ->  Parallel Seq Scan on procrastinate_jobs jobs  (cost=0.00..4534.28 rows=89598 width=191) (actual time=0.015..18.812 rows=76164 loops=2)
                           Filter: ((status = 'todo'::procrastinate_job_status) AND ((scheduled_at IS NULL) OR (scheduled_at <= now())))
                           Rows Removed by Filter: 10
                     ->  Parallel Hash  (cost=3754.97..3754.97 rows=63465 width=40) (actual time=12.739..12.740 rows=76174 loops=2)
                           Buckets: 262144  Batches: 1  Memory Usage: 2048kB
                           ->  Parallel Index Only Scan using procrastinate_jobs_id_lock_idx on procrastinate_jobs earlier_jobs  (cost=0.42..3754.97 rows=63465 width=40) (actual time=0.024..9.350 rows=76174 loops=2)
                                 Heap Fetches: 53
 Planning Time: 0.382 ms
 Execution Time: 71.563 ms
(20 rows)
```

So, my index isn't used at all, but the performance does not seem to degrade anymore with more jobs. I will delete my added index.

Reply:
Good for me.